### PR TITLE
Support for HTML media type

### PR DESCRIPTION
**Pull request by pichiste**
Hi there, 

This adds support for individual HTML files as a media type, and is aimed at the use case of interactive artworks. Currently artists are using/hijacking the SVG component to embed interactive pieces. Having an HTML media type would simplify the process and make it clearer to the user. It's a close copy of the SVG component by @andrevenancio.

### CODE REVIEW

**Comment by andrevenancio** [comment]
Nice one I had this on my todo list. But needs @crzypatchwork to give it a green light. Because this is Pandora's box open 😂🙌🏻❤️

**Comment by pichiste** [comment]
Pandora's box indeed!

**Comment by andrevenancio** [comment]
Give me some time until Brazil wakes up and I'll ask if I can merge this. It seems a better option that SVG embeds.

**Comment by pichiste** [comment]
Thanks!! 🙏  I don't think there are any new security concerns, more just a usability thing really.

**Comment by mrdoob** [comment]
I was worried yesterday that people were able to embed other websites inside SVGs.
But [I'm glad to see](https://twitter.com/quasimondo/status/1372859807607164930
) that removing `allow-same-origin` solved that.

Security-wise, I'm now more confident about supporting `html` 🤓

**Comment by mrdoob** [comment]
Also, I don't think this is too different from what SVGs `foreignObject` already allows the developer to do.
So I don't see it much as opening pandora's box but rather to stop abusing SVGs 😁

**Comment by spite** [comment]
Related: what do we want to do about permissions that require https://developer.mozilla.org/en-US/docs/Web/HTTP/Feature_Policy (e.g. Camera, Orientation, XR)?

Edit: more concretely, i think that if those were allowed, they should probably be disabled on the main feed and only allowed in the objkt page.

**Comment by mrdoob** [comment]
@spite One step at a time? 😁

**Comment by spite** [comment]
@mrdoob of course. but i don't think ignoring the overall security is the right call if it can be considered at the same time. specially if one thing doesn't exclude the other. better to be safe than sorry

**Comment by mrdoob** [comment]
As far as I understand...
None of these features are allowed unless we add, say... `allow="camera; fullscreen"` to the iframe.

**Comment by nickdima** [comment]
Together with security we should consider also the immutability aspect. What if you call an external script in the HTML page that's later modified? 
Even if hicetnunc will have some security measures in place, once these HTML NFTs are on the blockchain you never know how other websites/apps will handle them.

**Comment by pichiste** [comment]
@spite @mrdoob Yup it does block the features by default, so they'd have to be selectively enabled with `allow`. I guess it would just be a matter of pairing down the feature list to one that makes sense at some point?

My wishlist 😅
```
accelerometer
autoplay
camera
fullscreen
gyroscope
magnetometer
microphone
xr-spatial-tracking (!!!)
```

Would love to just see basic HTML working first though!

**Comment by mrdoob** [comment]
@nickdima The iframe only has `sandbox="allow-scripts"`. It can't access external urls.

**Comment by nickdima** [comment]
> @nickdima The iframe only has `sandbox="allow-scripts"`. It can't access external urls.

Yeah what I was thinking is that the responsibility is given to whoever loads that HTML NFT and mistakes might be made. Imagine a new marketplace loads this NFTs and they "forget" to sandbox the iframe.

**Comment by pichiste** [comment]
@nickdima @mrdoob  Hm just tested and it looks like `sandbox` can't prevent the iframe from requesting random web resources once scripts are enabled.

I'm not convinced that external resources should be prevented though... If someone opens the direct link, they are at the same risk as visiting any webpage, no?

If we really need it to be standalone, a few possible solutions:
- Injecting a Content Security meta tag upon upload, as suggested [here](https://github.com/hicetnunc2000/hicetnunc/issues/14#issuecomment-800555041).
- Wrapping the HTML file in an iframe upon upload, so you are already getting a sandboxed iframe.
- Upon upload, fetching all external resources and injecting them inline into the main doc.

These would happen on the frontend though so not sure how secure they really are...

With regard to this PR though, the SVG element works the same as this, so it's not introducing a new vulnerability.

**Comment by pichiste** [comment]
> > @nickdima The iframe only has `sandbox="allow-scripts"`. It can't access external urls.
> 
> Yeah what I was thinking is that the responsibility is given to whoever loads that HTML NFT and mistakes might be made. Imagine a new marketplace loads this NFTs and they "forget" to sandbox the iframe.

Fair point!

**Comment by pichiste** [comment]
Testing this out, the CSP meta tag feels like the best solution. Simply injecting this into `<head>` upon upload:
```html
<meta http-equiv="Content-Security-Policy" content="default-src 'self';">
```

Thoughts?

**Comment by mrdoob** [comment]
@pichiste Is it possible to remove the meta tag afterwards with JS? 😬

**Comment by mrdoob** [comment]
Is so... maybe double iframe would be the easiest/safest...

**Comment by pichiste** [comment]
@mrdoob Good call!

I just tested and it looks like fortunately the CSP policy takes effect on load, so it doesn't matter if it gets removed in the js. 

Another thing about the double iframes is that things like the feature policy mentioned above would potentially need to be baked into the inner iframe, though maybe its safer in some ways?

**Comment by pichiste** [comment]
Ok, I've gone ahead and implemented the CSP meta tag solution. Basically [these utility functions](https://github.com/pichiste/hicetnunc/blob/edfd0124df7a1ed6530e1afae700dcc54e5ebfae/src/utils/html.js) are used to inject the following meta tag into both the [HTML preview](https://github.com/pichiste/hicetnunc/blob/edfd0124df7a1ed6530e1afae700dcc54e5ebfae/src/components/media-types/html/index.js#L12) and [uploaded HTML file](https://github.com/pichiste/hicetnunc/blob/edfd0124df7a1ed6530e1afae700dcc54e5ebfae/src/pages/mint/index.js#L45):

```html
<meta http-equiv="Content-Security-Policy" content="default-src 'self'; script-src 'unsafe-inline'; style-src 'unsafe-inline'">
```

I think that covers external sources concern.

Anything else?

**Comment by mrdoob** [comment]
Haha!

I thought you were just going to do `buffer = '<meta http-equiv="Content-Security-Policy">' + buffer` 😁

Beautiful!

File: src/utils/html.js
Comment (mrdoob) [inline src/utils/html.js]: Maybe also check if the html already has a `Content-Security-Policy` meta and delete it?
What happens if there are two?

File: src/components/media-types/index.js
Comment (andrevenancio) [inline src/components/media-types/index.js]: check this: https://github.com/hicetnunc2000/hicetnunc/blob/main/src/components/media-types/index.js#L37

I've injected some properties on the SVG component that allows me to inject 2 props, owner and viewer in case you want the content of your html hashed to whoever is seeing your creation.

Comment (andrevenancio) [reply src/components/media-types/index.js]: (Only injected on the objkt detail page)

File: src/utils/html.js
Comment (pichiste) [reply src/utils/html.js]: Good call, thanks! Fixed here: https://github.com/hicetnunc2000/hicetnunc/pull/189/commits/2ce5c0ec337f384397b9ac15f6b6456551d70ab1

File: src/components/media-types/index.js
Comment (pichiste) [reply src/components/media-types/index.js]: Oh nice! The hashing is a really great idea. Went ahead and added it: https://github.com/hicetnunc2000/hicetnunc/pull/189/commits/2042b1dfaa1919b9403f706bda4236fa14759b88

Comment (pichiste) [reply src/components/media-types/index.js]: Ah we should probably add the CSP meta tag to SVG content too, no? Before there was a `sandbox-svg.html` file but where would be the place now?

Comment (pichiste) [reply src/components/media-types/index.js]: Sorry I see what changed now. Still not sure of best way to add the meta tag to the SVG iframe, but maybe that can be a separate issue to this?

Hope its ok I made a small change to `VectorComponent` because it wasn't displaying properly in preview mode with the author and viewer query params in the src since its a data URI at that point (unless I'm misunderstanding something?):
https://github.com/pichiste/hicetnunc/commit/a02e60b69bb126683faf9809633589e3fcafd023

Comment (pichiste) [reply src/components/media-types/index.js]: Resolving for now but let me know if anything needs to change, thanks!!

Comment (andrevenancio) [reply src/components/media-types/index.js]: The hash wasn't available in preview mode because I wasn't putting in a url to a file. But if this fixes it then great! I'm gonna give this a test tonight. Can you drop a html template on the templates folder now on main? Just so I can use to test the file upload

Comment (pichiste) [reply src/components/media-types/index.js]: Cool, just added an html template (https://github.com/hicetnunc2000/hicetnunc/pull/189/commits/9c5e9b4bca18b0c2c9b2199c42bdd81fa37920a2)!

A heads up, both svg and html templates reference the query param `creator` but it's being set as `author` in `VectorComponent` and `HTMLComponent`. Left it that way since I didn't know if it's being used elsewhere.

**Review by andrevenancio — APPROVED** [review]

**Comment by andrevenancio** [comment]
merged this to develop. but will not release just yet. need to double check with @crzypatchwork

**Comment by pichiste** [comment]
Awesome thanks!! I guess one last thing before releasing would be to make sure the author/creator query param name matches between the templates and vector/html iframe embeds. 

Excited to see this live!

**Pull request merged**
